### PR TITLE
test(connector): add test cases for debezium json test

### LINKED ISSUE #8469: bug(cdc-connector): NULL date is converted to 1970-01-01
### Describe the bug

On Postgres:

image

On RisignWave:

image

### To Reproduce

As shown above.

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by WillyKidd**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- introduced testcontainers (mysql and postgres) to test debezium cdc connectors
- added test cases for the debezium json parser
- modified the debezium json parser
  - added conversion from i16 to bool (since debezium connector converts bool in mysql/pg to i16)
  - added conversion from i32 to date (since debezium connector converts date in mysql/pg to i32)
  - added conversion from i64 to time (since debezium connector converts time in mysql/pg to i64)
  - added overflow detection for f32 (the parser would otherwise convert to `inf` without a warning)

[debezium mysql data type mappings](https://debezium.io/documentation/reference/stable/connectors/mysql.html#mysql-data-types)
[debezium pg data type mappings](https://debezium.io/documentation/reference/stable/connectors/postgresql.html#postgresql-data-types)

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [x] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Release note

</details>

fix #8469

### CODE REVIEW

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+                        + "O_BOOL BOOLEAN, "
+                        + "O_TINY TINYINT, "
+                        + "O_INT INT, "
+                        + "O_REAL REAL, "
+                        + "O_DOUBLE DOUBLE, "
+                        + "O_DECIMAL DECIMAL(15, 2), "
+                        + "O_CHAR CHAR(15), "
+                        + "O_DATE DATE, "
+                        + "O_TIME TIME, "
+                        + "O_TIMESTAMP TIMESTAMP, "
```
Comment (tabVersion) [inline connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java:145]: please cover corner cases
for date: `1000-01-01` and `9999-12-31`
for time: `-838:59:59` and `838:59:59`
for datetime: `1000-01-01 00:00:00` and `9999-12-31 23:59:59`
for timestamp: `'1970-01-01 00:00:01' UTC` and `'2038-01-09 03:14:07' UTC` (and other timezones)

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    public void getTestJson() throws InterruptedException, SQLException {
+        Connection connection = SourceTestClient.connect(mysql);
+        String query =
+                "CREATE TABLE IF NOT EXISTS orders ("
+                        + "O_KEY BIGINT NOT NULL, "
+                        + "O_BOOL BOOLEAN, "
+                        + "O_TINY TINYINT, "
+                        + "O_INT INT, "
+                        + "O_REAL REAL, "
+                        + "O_DOUBLE DOUBLE, "
```
Comment (tabVersion) [inline connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java:140]: please cover overflow cases

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java (hunk 1 above)
Comment (WillyKidd) [reply connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java:145]: thanks for your comments bohan! i am wondering what is the expected behavior when converting mysql time to postgres time (since risingwave is postgres compatible)...

https://support.huaweicloud.com/intl/en-us/productdesc-drs/drs_01_0324.html

> TIME values supported by MySQL range from '-838:59:59' to '838:59:59'. For details, see the [official MySQL documentation](https://dev.mysql.com/doc/refman/5.7/en/time.html). For PostgreSQL, the minimum value of the TIME type is 00:00:00 and the maximum value is 24:00:00. In MySQL, if a value of the TIME type is less than 00:00:00 or greater than 24:00:00, DRS will convert it to 00:00:00.

I guess we can either let it overflow, or convert it to 00:00:00 as above...?

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java (hunk 2 above)
Comment (WillyKidd) [reply connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java:140]: as for the overflow cases...
when inserting data into mysql, if there is overflow, mysql will throw an error:
> Data truncation: Out of range value for column 'O_INT' at row 1
so it's possibly unlikely that debezium will generate overflowing values...?

I can also try to modify the json generated by debezium and see if the parser catches the error

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java (hunk 1 above)
Comment (StrikeW) [reply connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java:145]: If the value is out of range of RisingWave's type, which means the data may have been corrupted in upstream. Thus we can report an out-of-range (e.g. underflow, overflow) error to log and skip this row.

Comment (tabVersion) [reply connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java:145]: Sure, I mean we should cover them in test cases and document the behavior.

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java (hunk 2 above)
Comment (tabVersion) [reply connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/MySQLSourceTest.java:140]: debezium is unlikely to deliver the corrupted data but we still may face it when the data comes from kafka.
Just mock some test cases and see if rw catches the error.

File: connector_node/risingwave-source-test/pom.xml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        <dependency>
+            <groupId>com.zaxxer</groupId>
+            <artifactId>HikariCP</artifactId>
+            <version>5.0.1</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>testcontainers</artifactId>
+            <version>1.17.6</version>
```
Comment (StrikeW) [inline connector_node/risingwave-source-test/pom.xml:37]: Extract the version into a property in the pom.xml

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.PostgreSQLContainer;
+import org.testcontainers.utility.MountableFile;
+
+public class PostgresSourceTest {
+    private static final Logger logger =
+            LoggerFactory.getLogger(PostgresSourceTest.class.getName());
+
+    private static final PostgreSQLContainer<?> pg =
+            new PostgreSQLContainer<>("postgres:12.3-alpine")
```
Comment (StrikeW) [inline connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java:32]: What does the `alpine` mean in the version number?

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.junit.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.PostgreSQLContainer;
+import org.testcontainers.utility.MountableFile;
+
+public class PostgresSourceTest {
+    private static final Logger logger =
```
Comment (StrikeW) [inline connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java:28]: Generally, the `static final` field should be all uppercase, e.g. `LOG`

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+            new SourceTestClient(
+                    Grpc.newChannelBuilder(
+                                    "localhost:" + ConnectorService.DEFAULT_PORT,
+                                    InsecureChannelCredentials.create())
+                            .build());
+
+    @BeforeClass
+    public static void init() {
+        // generate orders.tbl test data
+        SourceTestClient.genOrdersTable(10000);
```
Comment (StrikeW) [inline connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java:52]: Would you generate two copy of test data for Mysql and PG?

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+            connectorServer.start();
+            logger.info("connector service started");
+            pg.withCopyFileToContainer(
+                    MountableFile.forClasspathResource("orders.tbl"), "/home/orders.tbl");
+            pg.start();
+            pg.withUsername("postgres")
+                    .execInContainer(
+                            "sh",
+                            "-c",
+                            "echo 'host replication postgres 172.17.0.1/32 trust' >> /var/lib/postgresql/data/pg_hba.conf");
```
Comment (StrikeW) [inline connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java:64]: What's the purpose of this command?

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/SourceTestClient.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            } else {
+                LOG.info("updated: " + statement.getUpdateCount());
+            }
+        } catch (SQLException e) {
+            LOG.warn("SQL Exception: {}", e.getMessage());
+        }
+        return resultSet;
+    }
+
+    protected static DataSource getDataSource(JdbcDatabaseContainer<?> container) {
```
Comment (StrikeW) [inline connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/SourceTestClient.java:69]: It is definitely unnecessary to build a brand new DataSource each time for every connection. You should put it in the class field.

**Review by StrikeW — COMMENTED** [review]
Please refine the java code to improve the quality a bit.

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/SourceTestClient.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+public class SourceTestClient {
+    static final Logger LOG = LoggerFactory.getLogger(SourceTestClient.class.getName());
+    private final ConnectorServiceGrpc.ConnectorServiceBlockingStub blockingStub;
+
+    public Properties sqlStmts = new Properties();
+
+    public SourceTestClient(Channel channel) {
+        blockingStub =
+                ConnectorServiceGrpc.newBlockingStub(channel).withMaxInboundMessageSize(16777216);
```
[the remark this answers is not in the document]
Comment (WillyKidd) [reply connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/SourceTestClient.java:34]: no special reason for this number as I randomly chose 16mb...
the default message size would be too small for the test when inserting 10_000 rows

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java (hunk 7 above)
Comment (WillyKidd) [reply connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java:64]: this command is to specify the clients that are permitted to replicate with the postgres, so that debezium can work properly. 
seems like currently testcontainers has no easy way for us to modify/override `pg_hba.conf` so I decided to use thi s command as a workaround..
https://debezium.io/documentation/reference/stable/connectors/postgresql.html#postgresql-host-replication-permissions

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java (hunk 6 above)
Comment (WillyKidd) [reply connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java:52]: yes it seems so...maybe I can first check the line number before deciding whether to write data or not :)

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java (hunk 4 above)
Comment (WillyKidd) [reply connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/PostgresSourceTest.java:32]: to be honest I randomly picked an image...
from stack overflow:
> Alpine is a much smaller version of Linux, it results in a smaller container than the full postgres image.
https://stackoverflow.com/questions/62333176/docker-difference-postgres12-from-postgres12-alpine

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/SourceTestClient.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+    private static String[] orderStatusArr = {"O", "F"};
+    private static String[] orderPriorityArr = {
+        "1-URGENT", "2-HIGH", "3-MEDIUM", "4-NOT SPECIFIED", "5-LOW"
+    };
+
+    // generates an orders table in class path if not exists
+    // data completely random
+    static void genOrdersTable(int numRows) {
+        String path =
+                PostgresSourceTest.class
```
[the remark this answers is not in the document]
Comment (WillyKidd) [reply connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/SourceTestClient.java:119]: yes it was a mistake and it can be confusing...changed it to `SourceTestClient`

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/SourceTestClient.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (StrikeW) [reply connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/SourceTestClient.java:34]: What do you mean by too small? Any errors if you don't modify this config?

**Comment by StrikeW** [comment]
@waruto210 Please take a look at the modification to the parser.

**Review by waruto210 — COMMENTED** [review]
LGTM

File: src/expr/src/vector_op/cast.rs [hunk 11]
```diff
@@ -49,11 +50,29 @@ const PARSE_ERROR_STR_TO_TIME: &str =
 const PARSE_ERROR_STR_TO_DATE: &str = "Can't cast string to date (expected format is YYYY-MM-DD)";
 const PARSE_ERROR_STR_TO_BYTEA: &str = "Invalid Bytea syntax";
 
+#[inline(always)]
+pub fn i32_to_date(d: i32) -> Result<NaiveDateWrapper> {
+    Ok(NaiveDateWrapper::new(
+        NaiveDate::from_num_days_from_ce_opt(d).unwrap() + Days::new(719_163),
```
Comment (StrikeW) [inline src/expr/src/vector_op/cast.rs:56]: Need comment for this function. The input integer should represent the number of days since the epoch (`1 January 1970`).

File: src/expr/src/vector_op/cast.rs [hunk 12]
```diff
@@ -49,11 +50,29 @@ const PARSE_ERROR_STR_TO_TIME: &str =
...
+    ))
+}
+
 #[inline(always)]
 pub fn str_to_date(elem: &str) -> Result<NaiveDateWrapper> {
     Ok(NaiveDateWrapper::new(parse_naive_date(elem)?))
 }
 
+#[inline(always)]
+pub fn i64_to_time(t: i64) -> Result<NaiveTimeWrapper> {
```
Comment (StrikeW) [inline src/expr/src/vector_op/cast.rs:66]: Need comments for this function. The input i64 is in ms or us unit?

File: src/connector/src/parser/common.rs [hunk 13]
```diff
@@ -18,19 +18,28 @@ use risingwave_common::array::{ListValue, StructValue};
...
+            BorrowedValue::Static(StaticNode::Bool(_)) => {
+                v.as_bool().ok_or_else(|| anyhow!("expect bool"))?.into()
+            }
+            // debezium converts bool to int, false -> 0, true -> 1, for mysql and postgres
+            BorrowedValue::Static(StaticNode::I64(i)) => match i {
+                0i64 => ScalarImpl::Bool(false),
+                1i64 => ScalarImpl::Bool(true),
+                _ => anyhow::bail!("expect bool, but found {v}"),
+            },
+            _ => anyhow::bail!("expect bool, but found {v}"),
```
Comment (xiangjinwu) [inline src/connector/src/parser/common.rs:41]: * When matching bool, it can be just `BorrowedValue::Static(StaticNode::Bool(b)) => (*b).into()`
* When matching int, avoid relying on `0` or `1` being an `I64` rather than `U64`. Prefer `as_` to cover all cases.

File: src/connector/src/parser/common.rs [hunk 14]
```diff
@@ -42,16 +51,43 @@ fn do_parse_simd_json_value(dtype: &DataType, v: &BorrowedValue<'_>) -> Result<S
...
         DataType::Int64 => ensure_int!(v, i64).into(),
-        DataType::Float32 => ScalarImpl::Float32((simd_json_ensure_float!(v, f32) as f32).into()),
-        DataType::Float64 => ScalarImpl::Float64((simd_json_ensure_float!(v, f64)).into()),
+        // when f32 overflows, the value is converted to `inf` which is inappropriate
+        DataType::Float32 => {
+            let scalar_val = ScalarImpl::Float32((simd_json_ensure_float!(v, f32) as f32).into());
+            if let ScalarImpl::Float32(f) = scalar_val {
+                if f.is_infinite() {
+                    anyhow::bail!("{v} is out of range for type f32");
+                }
```
Comment (xiangjinwu) [inline src/connector/src/parser/common.rs:60]: FYI this is okay only in the context of json, where the original value cannot be infinity.

File: src/connector/src/parser/common.rs [hunk 15]
```diff
@@ -42,16 +51,43 @@ fn do_parse_simd_json_value(dtype: &DataType, v: &BorrowedValue<'_>) -> Result<S
...
+            }
+            scalar_val
+        }
+        DataType::Float64 => {
+            let scalar_val = ScalarImpl::Float64((simd_json_ensure_float!(v, f64)).into());
+            if let ScalarImpl::Float64(f) = scalar_val {
+                if f.is_infinite() {
+                    anyhow::bail!("{v} is out of range for type f64");
+                }
```
Comment (xiangjinwu) [inline src/connector/src/parser/common.rs:69]: When does this happen?

File: src/connector/src/parser/common.rs [hunk 16]
```diff
@@ -42,16 +51,43 @@ fn do_parse_simd_json_value(dtype: &DataType, v: &BorrowedValue<'_>) -> Result<S
...
         DataType::Decimal => Decimal::from_f64(simd_json_ensure_float!(v, Decimal))
             .ok_or_else(|| anyhow!("expect decimal"))?
             .into(),
         DataType::Varchar => ensure_str!(v, "varchar").to_string().into(),
         DataType::Bytea => ensure_str!(v, "bytea").to_string().into(),
-        DataType::Date => str_to_date(ensure_str!(v, "date"))?.into(),
-        DataType::Time => str_to_time(ensure_str!(v, "time"))?.into(),
+        // debezium converts date to i32 for mysql and postgres
+        DataType::Date => match v {
+            BorrowedValue::String(_) => str_to_date(ensure_str!(v, "date"))?.into(),
```
Comment (xiangjinwu) [inline src/connector/src/parser/common.rs:81]: Same as the bool case above, you are already guaranteed a string and there is no need to `ensure` again:
```
BorrowedValue::String(s) => str_to_date(s)?.into()
```

File: src/connector/src/parser/common.rs [hunk 17]
```diff
@@ -24,42 +26,67 @@ use risingwave_expr::vector_op::cast::{
...
+        DataType::Date => match v {
+            BorrowedValue::String(s) => str_to_date(s)?.into(),
+            BorrowedValue::Static(_) => {
+                NaiveDateWrapper::with_days_since_unix_epoch(ensure_i32!(v, i32))?.into()
+            }
+            _ => anyhow::bail!("expect date, but found {v}"),
+        },
+        // debezium converts time to i64 for mysql and postgres
+        DataType::Time => match v {
+            BorrowedValue::String(_) => str_to_time(ensure_str!(v, "time"))?.into(),
```
Comment (xiangjinwu) [inline src/connector/src/parser/common.rs:73]: Same as above. No need to ensure str again.

File: src/connector/src/parser/common.rs [hunk 18]
```diff
@@ -24,42 +26,67 @@ use risingwave_expr::vector_op::cast::{
...
+            BorrowedValue::Static(_) => NaiveTimeWrapper::with_milli(
+                ensure_i64!(v, i64)
+                    .try_into()
+                    .map_err(|_| anyhow!("cannot cast i64 to time, value out of range"))?,
+            )?
+            .into(),
```
Comment (xiangjinwu) [inline src/connector/src/parser/common.rs:79]: Is it always `milli`?

https://debezium.io/documentation/reference/stable/connectors/mysql.html#mysql-temporal-types
https://debezium.io/documentation/reference/stable/connectors/postgresql.html#postgresql-basic-types

According to the docs, `time.precision.mode=connect` uses `milli` but `time.precision.mode=adaptive_time_microseconds` uses `micro`.

**Review by xiangjinwu — APPROVED** [review]
rest lgtm regarding the parser updates

File: src/common/src/types/chrono_wrapper.rs [hunk 19]
```diff
@@ -184,6 +184,15 @@ impl NaiveDateWrapper {
         ))
     }
 
+    pub fn with_days_since_unix_epoch(days: i32) -> Result<Self> {
+        Ok(NaiveDateWrapper::new(
+            NaiveDate::from_num_days_from_ce_opt(days)
+                .ok_or_else(|| InvalidParamsError::date(days))?
+                .checked_add_days(Days::new(UNIX_EPOCH_DAYS.try_into().unwrap()))
```
Comment (StrikeW) [inline src/common/src/types/chrono_wrapper.rs:191]: We should not unwrap, which may cause panic.

File: connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/SourceTestClient.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (WillyKidd) [reply connector_node/risingwave-source-test/src/test/java/com/risingwave/connector/SourceTestClient.java]: it doesnt seem necessary..deleted :)

File: src/connector/src/parser/common.rs (hunk 15 above)
Comment (WillyKidd) [reply src/connector/src/parser/common.rs]: deleted :)

**Review by StrikeW — APPROVED** [review]
LGTM!
